python/pyspark/sql/tests/test_udtf.py

```diff
@@ -1795,6 +1796,65 @@ def terminate(self):
                     assertSchemaEqual(df.schema, StructType().add("col1", IntegerType()))
                     assertDataFrameEqual(df, [Row(col1=10), Row(col1=100)])
 
+    def test_udtf_with_named_arguments(self):
+        @udtf(returnType="a: int")
+        class TestUDTF:
+            def eval(self, a, b):
+                yield a,
+
+        self.spark.udtf.register("test_udtf", TestUDTF)
+
+        for i, df in enumerate(
+            [
+                self.spark.sql("SELECT * FROM test_udtf(a=>10, b=>'x')"),
+                self.spark.sql("SELECT * FROM test_udtf(b=>'x', a=>10)"),
+            ]
+        ):
+            with self.subTest(query_no=i):
+                assertDataFrameEqual(df, [Row(a=10)])
+
+    def test_udtf_with_kwargs(self):
+        @udtf(returnType="a: int, b: string")
+        class TestUDTF:
+            def eval(self, **kwargs):
+                yield kwargs["a"], kwargs["b"]
+
+        self.spark.udtf.register("test_udtf", TestUDTF)
+
+        for i, df in enumerate(
+            [
+                self.spark.sql("SELECT * FROM test_udtf(a=>10, b=>'x')"),
+                self.spark.sql("SELECT * FROM test_udtf(b=>'x', a=>10)"),
```

Comment:
What would be the error message if the named argument is used incorrectly? For example
1. duplicated input argument names: a => 10, a => 10
2. non-existing argument name: c => 10
3. incorrect combination of positional and named arguments: test_udtf(a => 10, 'x')

I am afraid that if we directly leverage Python's kwargs, the error messages wouldn't be as user-friendly as the SQL function ones.

Reply:
That's a good point. So far just rely on the Python's error.
@dtenedor What's the error message like when applying name arguments with the above cases to other functions?

Reply:
Yeah I believe @learningchess2003 added these checks in [1]. They are currently in the `FunctionBuilderBase.scala` file in [2]. If we want to reuse those checks, we could be consistent between error messages for Python UDTFs and other Spark functions.

[1] https://github.com/apache/spark/pull/42020
[2] https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/FunctionBuilderBase.scala#L107-L128

Reply:
Updated to raise the following errors:

> 1. duplicated input argument names: a => 10, a => 10

It will be checked in the analysis phase and an error with the error class `DUPLICATE_ROUTINE_PARAMETER_ASSIGNMENT.DOUBLE_NAMED_ARGUMENT_REFERENCE` will be raised.

> 2. non-existing argument name: c => 10

It will be handled in Python runtime and an error will be raised.

```
...PySparkRuntimeError: [UDTF_EXEC_ERROR] User defined table function encountered an error in the 'eval' method: eval() got an unexpected keyword argument 'c'
```

> 3. incorrect combination of positional and named arguments: test_udtf(a => 10, 'x')

It will be checked in the analysis phase and an error with the error class `UNEXPECTED_POSITIONAL_ARGUMENT` will be raised.

Reply:
LGTM!